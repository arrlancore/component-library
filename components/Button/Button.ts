import styled from '@emotion/styled';
import isPropValid from '@emotion/is-prop-valid';
import {
  isObjectEmpty,
  spacing,
  fontSizes,
  ITheme,
  theme as defaultTheme,
} from '../../utils';

import { Theme } from '@storybook/theming';

type buttonPropSize ={
  [key: string]: any
}
const buttonSizeProps: buttonPropSize = {
  small: {
    fontSize: fontSizes['xsmall'],
    padding: `${spacing['xsmall']} ${spacing['small']}`,
  },
  medium: {
    fontSize: fontSizes['small'],
    padding: `${spacing['small']} ${spacing['medium']}`,
  },
  large: {
    fontSize: fontSizes['medium'],
    padding: `${spacing['medium']} ${spacing['large']}`,
  },
};

type PropsByVariant = {
  variant: string,
  color: string,
  theme: ITheme,
}

const getPropsByVariant = ({ variant, color, theme }: PropsByVariant) => {
  const colorInPalette = theme.palette[color];

  const defaultOutlineVariantProps = {
    main: {
      border: `1px solid ${theme.palette.common.black}`,
      backgroundColor: theme.palette.common.white,
      color: theme.palette.common.black,
    },
    hover: {
      border: `1px solid ${theme.palette.common.black}`,
      backgroundColor: theme.palette.common.white,
      color: theme.palette.common.black,
    },
  };

  const outlineVariantPropsByPalette = colorInPalette && {
    main: {
      border: `1px solid ${colorInPalette.main}`,
      backgroundColor: theme.palette.common.white,
      color: colorInPalette.main,
    },
    hover: {
      border: `1px solid ${colorInPalette.light}`,
      backgroundColor: theme.palette.common.white,
      color: colorInPalette.light,
    },
  };

  const defaultSolidVariantProps = {
    main: {
      border: `1px solid ${theme.palette.grey[100]}`,
      backgroundColor: theme.palette.grey[100],
      color: theme.palette.common.black,
    },
    hover: {
      border: `1px solid ${theme.palette.grey[200]}`,
      backgroundColor: theme.palette.grey[200],
    },
  };

  const solidVariantPropsByPalette = colorInPalette && {
    main: {
      border: `1px solid ${colorInPalette.main}`,
      backgroundColor: colorInPalette.main,
      color: colorInPalette.contrastText,
    },
    hover: {
      border: `1px solid ${colorInPalette.light}`,
      backgroundColor: colorInPalette.light,
    },
  };

  type variantType = {
    [key: string]: any
  }

  const variants: variantType = {
    outline: colorInPalette
      ? outlineVariantPropsByPalette
      : defaultOutlineVariantProps,
    solid: colorInPalette
      ? solidVariantPropsByPalette
      : defaultSolidVariantProps,
  };

  return variants[variant] || variants.solid;
};

export interface IButtonProps {
  color?: string,
  size?: string,
  variant?: string,
  enableElevation?: boolean,
  disabled?: boolean,
  theme?: any,
  onClick?: any
}

const StyledButton = ({
  color,
  size,
  variant,
  enableElevation,
  disabled,
  theme,
}: IButtonProps) => {
  if (isObjectEmpty(theme)) {
    theme = defaultTheme;
  }

  const fontSizeBySize = buttonSizeProps[size]?.fontSize;
  const paddingBySize = buttonSizeProps[size]?.padding;

  const propsByVariant = getPropsByVariant({ variant, theme, color });

  return {
    fontWeight: 500,
    cursor: 'pointer',
    opacity: disabled && 0.7,
    transition: 'all 0.3s linear',
    padding: buttonSizeProps.medium.padding,
    fontSize: buttonSizeProps.medium.fontSize,
    borderRadius: theme.shape.borderRadius,
    fontFamily: theme.typography.fontFamily,
    boxShadow: enableElevation && theme.shadows[1],
    ...(propsByVariant && propsByVariant.main),
    ...(paddingBySize && { padding: paddingBySize }),
    ...(fontSizeBySize && { fontSize: fontSizeBySize }),
    '&:hover': !disabled && {
      boxShadow: enableElevation && theme.shadows[2],
      ...(propsByVariant && propsByVariant.hover),
    },
  };
};

const IGNORED_PROPS = ['color'];

const buttonConfig = {
  shouldForwardProp: (prop: any) =>
    isPropValid(prop) && !IGNORED_PROPS.includes(prop),
};

export const Button = styled('button', buttonConfig)(StyledButton);

