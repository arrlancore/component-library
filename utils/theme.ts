import { transparentize } from 'polished';
import { spacing } from './units';

const white = transparentize(0.75, '#fff');
const white75 = transparentize(0.5, '#fff');
const white50 = transparentize(0.25, '#fff');
const black = '#111';
const yellow = '#F2CC23';
const yellowDarker = '#E3BF22';
const green = '#17A167';
const red = '#F96854';
const red75 = '#F96854';
const grey = '#B9B8B9';
// background
const blackDark = '#2D2F33';
const blackDarker = '#282A2E';
const blackDarkest = '#202225';
// button
const greyDark = '#494B50';
const grey10 = '#B9B8B9';
const redError = '#DB524E';

const palette = {
  common: {
    white,
    white75,
    white50,
    black,
    yellow,
    yellowDarker,
    green,
    red,
    red75,
    grey,
    blackDark,
    blackDarker,
    blackDarkest,
    greyDark,
    grey10,
    redError,
  },
  primary: {
    main: '#0070F3',
    light: '#146DD6',
    contrastText: white,
  },
  error: {
    main: '#A51C30',
    light: '#A7333F',
    contrastText: white,
  },
  grey: {
    100: '#EAEAEA',
    200: '#C9C5C5',
    300: '#888',
    400: '#666',
  },
};

const shadows = {
  0: 'none',
  1: '0px 5px 10px rgba(0, 0, 0, 0.12)',
  2: '0px 8px 30px rgba(0, 0, 0, 0.24)',
};

const typography = {
  fontFamily:
    "Inter, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Ubuntu, 'Helvetica Neue', sans-serif",
};

const shape = {
  borderRadius: spacing['xxsmall'],
};

type palleteTypes = {
  [key: string]: any
}

export interface ITheme {
  palette: palleteTypes,
  shadows: object,
  typography: object,
  shape: object
}

export const theme: ITheme = {
  palette,
  shadows,
  typography,
  shape,
};
