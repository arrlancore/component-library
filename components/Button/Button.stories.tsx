import * as React from 'react';
import { ThemeProvider } from 'emotion-theming';
import { boolean, text } from '@storybook/addon-knobs';
import decorators from '../../.storybook/utils';
import { theme as defaultTheme } from '../../utils';
import { Button } from '../';

export default {
  title: 'Components/Button',
  decorators,
};

export const SolidButton = () => {
  return <Button> Solid Button </Button>;
};

export const SolidButtonWithProps = () => {
  return (
    <Button
      color={text('color', '')}
      size={text('size', '')}
      disabled={boolean('disabled', false)}
      enableElevation={boolean('enableElevation', false)}
    >
      Solid Button
    </Button>
  );
};

export const OutlineButtonWithProps = () => {
  return (
    <Button
      variant="outline"
      size={text('size', '')}
      disabled={boolean('disabled', false)}
      enableElevation={boolean('enableElevation', false)}
    >
      Outline Button
    </Button>
  );
};

const theme = {
  ...defaultTheme,
  palette: {
    ...defaultTheme.palette,
    orange: {
      main: '#F34706',
      light: '#FF5964',
    },
  },
};

export const SolidButtonWithTheme = () => {
  return (
    <ThemeProvider theme={theme}>
      <Button
        variant={text('variant', 'solid')}
        color={text('color', 'orange')}
        size={text('size', 'medium')}
        disabled={boolean('disabled', false)}
        enableElevation={boolean('enableElevation', false)}
      >
        Outline Button
      </Button>
    </ThemeProvider>
  );
};

