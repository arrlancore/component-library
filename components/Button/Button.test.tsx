import * as React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import { Button } from './Button';
import serializer, { matchers } from 'jest-emotion'
import { fontSizes, spacing, theme } from '../../utils';

expect.addSnapshotSerializer(serializer)
expect.extend(matchers)

describe('<Button />', () => {
  it('should render the default button with a text', () => {
    render(<Button>Default Button</Button>)
    expect(screen.getByText('Default Button')).toBeInTheDocument()
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should occured some click event', () => {
    const onClick = jest.fn()
    render(<Button onClick={onClick}>Default Button</Button>)
    // click one time
    userEvent.click(screen.getByText('Default Button'))
    expect(onClick).toHaveBeenCalledTimes(1)

    // click two times
    userEvent.click(screen.getByText('Default Button'))
    expect(onClick).toHaveBeenCalledTimes(2)
  });

  it('should render with size small', () => {
    render(<Button size="small">With Size Props Small</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule('font-size', fontSizes['xsmall'])
    expect(screen.getByRole('button')).toHaveStyleRule('padding', `${spacing['xsmall']} ${spacing['small']}`)
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should render with size medium', () => {
    render(<Button size="medium">With Size Props Medium</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule('font-size', fontSizes['small'])
    expect(screen.getByRole('button')).toHaveStyleRule('padding', `${spacing['small']} ${spacing['medium']}`)
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should render with variant outline', () => {
    render(<Button size="medium" variant="outline">With variant</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule('border', `1px solid ${theme.palette.common.black}`)
    expect(screen.getByRole('button')).toHaveStyleRule('color', theme.palette.common.black)
    expect(screen.getByRole('button')).toHaveStyleRule('background-color', theme.palette.common.white)
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should render with variant outline & primary color', () => {
    render(<Button size="medium" color="primary" variant="outline">With Variant & color</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule('border', `1px solid ${theme.palette.primary.main}`)
    expect(screen.getByRole('button')).toHaveStyleRule('color', theme.palette.primary.main)
    expect(screen.getByRole('button')).toHaveStyleRule('background-color', theme.palette.common.white)
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should render with disabled attribute & cannot be click', () => {
    const onClick = jest.fn()
    render(<Button onClick={onClick} size="medium" disabled>With disabled</Button>)
    expect(screen.getByRole('button')).toBeDisabled()
    userEvent.click(screen.getByRole('button'))
    // cannot be clicked when disabled
    expect(onClick).not.toHaveBeenCalled()
    expect(screen.getByRole('button')).toMatchSnapshot()
  });

  it('should render with enable elevation', () => {
    render(<Button size="medium" enableElevation>With disabled</Button>)
    expect(screen.getByRole('button')).toHaveStyleRule('box-shadow', theme.shadows[1])
    expect(screen.getByRole('button')).toMatchSnapshot()
  });
});