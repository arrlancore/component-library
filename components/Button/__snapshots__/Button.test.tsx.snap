// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the default button with a text 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.5rem 0.75rem;
  font-size: 0.79rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #EAEAEA;
  background-color: #EAEAEA;
  color: #111;
}

.emotion-0:hover {
  border: 1px solid #C9C5C5;
  background-color: #C9C5C5;
}

<button
  class="emotion-0"
>
  Default Button
</button>
`;

exports[`<Button /> should render with disabled attribute & cannot be click 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.75rem 1.25rem;
  font-size: 0.889rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #EAEAEA;
  background-color: #EAEAEA;
  color: #111;
}

<button
  class="emotion-0"
  disabled=""
>
  With disabled
</button>
`;

exports[`<Button /> should render with enable elevation 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.75rem 1.25rem;
  font-size: 0.889rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  box-shadow: 0px 5px 10px rgba(0,0,0,0.12);
  border: 1px solid #EAEAEA;
  background-color: #EAEAEA;
  color: #111;
}

.emotion-0:hover {
  box-shadow: 0px 8px 30px rgba(0,0,0,0.24);
  border: 1px solid #C9C5C5;
  background-color: #C9C5C5;
}

<button
  class="emotion-0"
>
  With disabled
</button>
`;

exports[`<Button /> should render with size medium 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.75rem 1.25rem;
  font-size: 0.889rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #EAEAEA;
  background-color: #EAEAEA;
  color: #111;
}

.emotion-0:hover {
  border: 1px solid #C9C5C5;
  background-color: #C9C5C5;
}

<button
  class="emotion-0"
>
  With Size Props Medium
</button>
`;

exports[`<Button /> should render with size small 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.5rem 0.75rem;
  font-size: 0.79rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #EAEAEA;
  background-color: #EAEAEA;
  color: #111;
}

.emotion-0:hover {
  border: 1px solid #C9C5C5;
  background-color: #C9C5C5;
}

<button
  class="emotion-0"
>
  With Size Props Small
</button>
`;

exports[`<Button /> should render with variant outline & primary color 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.75rem 1.25rem;
  font-size: 0.889rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #0070F3;
  background-color: #fff;
  color: #0070F3;
}

.emotion-0:hover {
  border: 1px solid #146DD6;
  background-color: #fff;
  color: #146DD6;
}

<button
  class="emotion-0"
>
  With Variant & color
</button>
`;

exports[`<Button /> should render with variant outline 1`] = `
.emotion-0 {
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  padding: 0.75rem 1.25rem;
  font-size: 0.889rem;
  border-radius: 0.25rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Ubuntu,'Helvetica Neue',sans-serif;
  border: 1px solid #111;
  background-color: #fff;
  color: #111;
}

.emotion-0:hover {
  border: 1px solid #111;
  background-color: #fff;
  color: #111;
}

<button
  class="emotion-0"
>
  With variant
</button>
`;
